### DISCUSSION TITLE
[ZK Stack] - Displaying deposit history with thirdweb accounts

### DISCUSSION THREAD

**Discussion by franjohn21** [zkSync-Community-Hub/zksync-developers#817] [ZK Stack] [answered]
### Team or Project

Treasure

### Module Affected

- [ ] System contracts (era-system-contracts)
- [X] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [X] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.75.0

### Operating System

- [ ] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

Hey folks --

I'm running into issues trying to display deposit/withdrawal history for a given address. I am seeing the `Deposit` transaction show up under an intermediary address when used with Thirdweb. 

### Expected Behavior

 I would ideally like the block explorer API to categorize this as a deposit transaction for the recipient (or sender). `0xb7d834d0c50b03982783dc53274c4fab761e04e5` eg. https://block-explorer.topaz.treasurescan.io/address/0xb7d834d0c50b03982783dc53274c4fab761e04e5/transfers?type=deposit&page=1&limit=10)
 

### Current Behavior

What i'm seeing is that the `Deposit` transaction is attributed to an intermediary address (vs the destination/sending address). Here's an example:

Scenario: deposit magic for address `0xb7d834d0c50b03982783dc53274c4fab761e04e5` from Sepolia to Treasure Topaz via Thirdweb.

* [Sepolia deposit](https://sepolia.etherscan.io/tx/0x63025383472a882d37a3a70c5aad578727b3388b51ee5a27b34534b840b6d605) for address `0xb7d834d0c50b03982783dc53274c4fab761e04e5`.
* Within the NewPriorityRequest event emitted I see a reference to `e528132f0f0b8cad2c4b241b032154526f10de17`
* The L2 Block explorer only shows the `Transfer` transaction for `0xba4d3e1053af06b227de842bca0bb2729b092c23` , not the deposit. [Link](https://block-explorer.topaz.treasurescan.io/address/0xba4d3e1053af06b227de842bca0bb2729b092c23/transfers?page=1&limit=10)
* The L2 block explorer categorizes the `Deposit` transaction under `0xe528132f0f0b8cad2c4b241b032154526f10de17` [Link](https://block-explorer.topaz.treasurescan.io/address/e528132f0f0b8cad2c4b241b032154526f10de17/transfers?type=deposit&page=1&limit=10)

Questions:
* can the block explorer API return the deposit transaction under the intended recipient (`ba4d3e1053af06b227de842bca0bb2729b092c23`) 
* can I somehow retrieve this intermediary address (`0xe528132f0f0b8cad2c4b241b032154526f10de17`) so I know where to find the deposit transactions? is it deterministic for a given wallet address/chain?

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
Thank you for detailing the issue you're encountering with displaying deposit and withdrawal history in the Block Explorer when using Thirdweb.

Regarding your questions:

1. **Displaying Deposit Transactions Under Intended Recipient/Sender**: The Block Explorer is designed to show transactions based on the actual on-chain data. If the deposit transaction is showing under an intermediary address, it suggests that this address is part of the transaction path, possibly as a contract handling the deposits before they reach the final recipient. Unfortunately, the Block Explorer cannot alter how transactions are displayed unless the underlying data it receives changes. This behavior is determined by how the smart contracts for the deposits are coded.

2. **Retrieving Intermediary Address**: If the intermediary address (`0xe528132f0f0b8cad2c4b241b032154526f10de17`) is consistently used in the transaction process, it might be a contract address used by Thirdweb for handling deposits. This address is likely deterministic if it's part of a deployed contract's design. However, without specific details on how Thirdweb manages these addresses, it's challenging to confirm if every transaction or wallet will use the same intermediary address. You might need to consult Thirdweb's documentation or support for clarity on this behavior.

For a more tailored solution, you might consider reaching out to Thirdweb to understand if there's a way they can modify the transaction flow or provide additional APIs that could help in correctly attributing transactions to the intended addresses directly in your Block Explorer setup.

For further discussions and support, you can also visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by franjohn21** [comment, marked as answer]
Alright I believe this is either due to `refundRecipient` or something else but I've figured out the address can be determined by the [`applyL1ToL2Alias` function ](https://github.com/matter-labs/era-contracts/blob/aafee035db892689df3f7afe4b89fd6467a39313/l2-contracts/src/utils.ts#L30)

**Comment by dutterbutter** [comment]
Hello @franjohn21 👋 

Following up here, can you provide clarity what the current issue / feature request is? I see you've discovered `applyL1ToL2Alias` function so just want to see if anything has changed. Thanks!

**Comment by dutterbutter** [comment]
Closing due to staleness, feel free to re-open if needed!
